### PR TITLE
feat(api): API to return total number of copyrights for a file

### LINKED ISSUE #2468: Copyright File-view
File Details (Copyright/email/url/Author, ECC, IPRA) Content, Count, finding, activated/deactivated, delete + User findings with  description, text finding, comment for both activated and deactivated. With all decision options

### PR DESCRIPTION
**Pull request by soham4abc**
## Description

This API returns the total number of copyrights for a UploadTree

## How to test
Send a `GET` request to `/file/upload/{UploadId}/item/{ItemId}/totalcopyrights`

##Screenshots
Screenshot from 2023-06-21 22-47-51

Fixes #2468

@shaheemazmalmmd Please do have a look as this was requested on priority basis.

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2488"></a>

### CODE REVIEW

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+  public function __construct($container)
+  {
+    parent::__construct($container);
+    $this->container = $container;
+    $this->clearingDao = $this->container->get('dao.clearing');
+    $this->licenseDao = $this->container->get('dao.license');
+    $this->uploadDao = $container->get('dao.upload');
+    $this->restHelper = $container->get('helper.restHelper');
+    $this->CopyrightHist = $this->restHelper->getPlugin('ajax-copyright-hist');
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:65]: Please rebase to resolve issues here.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+   * @param  ResponseHelper         $response
+   * @param  array                  $args
+   * @return ResponseHelper
+   */
+  public function getTotalFileCopyrights($request, $response, $args)
+  {
+    $uploadPk = $args["id"];
+    $uploadTreeId = $args["itemId"];
+    $agentId = $this->CopyrightHist->getAgentId($uploadPk, 'copyright_ars');
+    $UploadTreeTableName = $this->uploadDao->getUploadtreeTableName($uploadPk);
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:81]:
```suggestion
    $uploadTreeTableName = $this->uploadDao->getUploadtreeTableName($uploadPk);
```

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    $returnVal = null;
+    try {
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:83]: Please remove `try/catch` as discussed.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      if (!$this->dbHelper->doesIdExist("upload", "upload_pk", $uploadPk)) {
+        $returnVal = new Info(404, "Upload does not exist", InfoType::ERROR);
+      } else if (!$this->dbHelper->doesIdExist($this->uploadDao->getUploadtreeTableName($uploadPk), "uploadtree_pk", $uploadTreeId)) {
+        $returnVal = new Info(404, "Item does not exist", InfoType::ERROR);
+      }
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:88]: Please move checks before accessing the variables.

File: src/www/ui/api/Controllers/CopyrightController.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      if ($returnVal !== null) {
+        return $response->withJson($returnVal->getArray(), $returnVal->getCode());
+      }
+      $returnVal = $this->CopyrightHist->getCopyrights($uploadPk, $uploadTreeId, $UploadTreeTableName, $agentId, 'statement', 'active', true);
+      $finalValue = array('total_copyrights' => $returnVal[2]);
+      return $response->withJson($finalValue, 200);
```
Comment (GMishx) [inline src/www/ui/api/Controllers/CopyrightController.php:94]: Please use more efficient ways:
```suggestion
      return $response->withJson(["total_copyrights" => $returnVal[2]], 200);
```

Comment (GMishx) [reply src/www/ui/api/Controllers/CopyrightController.php:94]: Even faster way is to move following lines to CopyrightDao as the function `getCopyrights()` also fetches other things from DB which are not required for this endpoint.

https://github.com/fossology/fossology/blob/fc89e1f3153d7c6eb088942ef6f0a2b44b58ade2/src/copyright/ui/ajax-copyright-hist.php#L288-L292

File: src/www/ui/api/documentation/openapi.yaml [hunk 6]
```diff
@@ -1728,6 +1728,56 @@ paths:
         default:
           $ref: '#/components/responses/defaultResponse'
 
+  /file/upload/{id}/item/{ItemId}/totalcopyrights:
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1731]: Please update path to be consistent with other endpoints.

File: src/www/ui/api/documentation/openapi.yaml [hunk 7]
```diff
@@ -3046,6 +3096,13 @@ components:
         uploadId:
           type: integer
           description: Upload ID to which the job belongs to.
+    getTotalFileCopyrights:
+      description: total copyrights of a file
+      type: object
+      properties:
+        total_copyrights:
+          type: integer
+          description: Total copyright number
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:3105]: Such simple response body can directly be embedded in the response of endpoint.

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by soham4abc** [comment]
@GMishx I was planning to add query parameter to this endpoint so that we can actually get the number of total copyrights for both active and inactive types.

Should i go for it?

**Comment by soham4abc** [comment]
> > @GMishx I was planning to add query parameter to this endpoint so that we can actually get the number of total copyrights for both active and inactive types.
> > Should i go for it?
> 
> Yes @soham4abc , it will be a good feature.

Done as per request sir.

Please do have a look!

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Pull request merged**